web3signer refactoring and test suite.

cheatfate opened the pull request.

cheatfate committed:
Initial commit.

cheatfate committed:
Refactor nimbus_signing_node to support Unix signals.

cheatfate committed:
Fix SN unable to close REST server properly.

cheatfate committed:
Fix `keys`, `deposit` and `validator_registration` endpoints issues.
Add getValidatorExitSignature() and getDepositMessageSignature() to validator_pool.

cheatfate committed:
Add /reload endpoint and implementation.
Fix signData to not cancel `timer`.
Fix validator_pool should clear attachedValidators table.

cheatfate committed:
Add test.

cheatfate committed:
Fix test issue.

cheatfate committed:
Fix AllTests.

cheatfate committed:
Post-rebase fixes.

cheatfate committed:
Recover non-windows tests.

cheatfate committed:
Diva protocol enhancement implementation.

cheatfate committed:
Fix AllTests.

zah commented on beacon_chain/nimbus_signing_node.nim:
```diff
@@ -231,143 +192,301 @@ proc installApiHandlers*(node: SigningNode) =
-  RestRouter.init(validate)
+  info "Initializaing validators", path = sn.config.validatorsDir()
+  var keysList: seq[string]
+  for keystore in listLoadableKeystores(sn.config, sn.keystoreCache):
```
Isn't this the same logic as the `reload` operation above? You can introduce a proc for this.

zah commented on beacon_chain/nimbus_signing_node.nim:
```diff
@@ -231,143 +192,301 @@ proc installApiHandlers*(node: SigningNode) =
+
+          if not(is_valid_merkle_branch(feeRecipientRoot, proof.merkleProofs,
+                                        log2trunc(proof.index),
+                                        get_subtree_index(proof.index),
```
You also need to check that `proof.index` has the expected value.
Otherwise, I'll be able to send a merkle proof for another field (e.g. the `graffitti` of the block) where I can put the expected fee recipient value in order to trick you.

zah commented on beacon_chain/spec/eth2_apis/rest_remote_signer_calls.nim:
```diff
@@ -70,132 +84,216 @@ proc signDataPlain*(identifier: ValidatorPubKey,
+          res = await client.signDataPlain(identifier, body,
+                                           restAcceptType = "application/json")
+          duration = Moment.now() - startSignMoment
+        nbc_remote_signer_time.observe(float(milliseconds(duration)) / 1000.0)
```
We have a naming convention where duration metrics use "_duration" as a suffix (although I can see this is not a new metric).

zah commented on beacon_chain/validators/validator_pool.nim:
```diff
@@ -435,6 +443,55 @@ proc signData(v: AttachedValidator,
   else:
     v.signWithDistributedKey(request)
 
+proc getFeeRecipientProof(blck: ForkedBeaconBlock | ForkedBlindedBeaconBlock |
```
The plan was to propose a more generalized solution where a list of generalized indices is specified in the remote keystore configuration, but we can deliver this in a follow-up PR.

cheatfate replied:
Right now the only change we do in remote keystore configuration is is type of remote keystore.

zah commented on beacon_chain/validators/validator_pool.nim:
```diff
@@ -413,17 +419,19 @@ proc signWithDistributedKey(v: AttachedValidator,
+    (client, info) = v.clients[0]
+    res = await client.signData(info.pubkey, deadline, 2, request)
+
+  if not(deadline.finished()): await cancelAndWait(deadline)
```
Why is it useful to cancel simple timers?
That's a genuine question: I'm not currently cancelling timers in my own code.

cheatfate replied:
Timers are not cheap right now, so it should be cancelled as well as other futures.

zah replied:
Can you clarify what's being saved by cancellation? I've taken a look at how `cancel` is implemented and it merely clears the `completion` function associated with the timer future (which in turn just calls `timerFut.complete()`).

cheatfate replied:
Yep, right now its not very effective, but as soon as we will have more effective implementation your code will start leaking timers and this leaks will be pretty hidden inside the code.

arnetheduck replied:
> clears the completion function

this btw has a tendency to save memory over time in general - closures have a tendency to keep futures alive longer than they need be, so being hygienic about clearing "obsolete" stuff also helps the GC

zah commented on beacon_chain/spec/eth2_apis/rest_remote_signer_calls.nim:
```diff
@@ -70,132 +84,216 @@ proc signDataPlain*(identifier: ValidatorPubKey,
+      var pending: seq[Future[void]]
+      if not(operationFut.finished()):
+        pending.add(operationFut.cancelAndWait())
+      await allFutures(pending)
```
`allFutures` can be called here with an empty sequence here. Is this a reasonable thing to do? Isn't this whole block unnecessarily complicated? It seems to me that you can just `await` the `operationFut` future when it's not finished.

zah commented on beacon_chain/spec/eth2_apis/rest_remote_signer_calls.nim:
```diff
@@ -70,132 +84,216 @@ proc signDataPlain*(identifier: ValidatorPubKey,
+                kind = resp.error.kind, reason = resp.error.message,
+                attempts_count = attemptsCount, attempt = attempt
+          lastError = Opt.some(resp.error)
+          inc(attempt)
```
A little bit of back-off would be appropriate here perhaps.

cheatfate replied:
I'm not sure i understand.

zah replied:
I meant waiting for an increasing number of miliseconds before each retry:

https://learn.microsoft.com/en-us/dotnet/architecture/microservices/implement-resilient-applications/implement-retries-exponential-backoff

cheatfate committed:
Address review comments.
Fix phase0 and altair blocks should not be signed when web3signer-diva protocol used.
Fix tests.

cheatfate committed:
Add back-off timeouts.

zah committed:
more traditional backoff logic

cheatfate commented:
@zah whole test suite missing distributive signing tests, because i dont know how to create distributed keys set.
